### PR TITLE
Add LifetimeParam and ConstParam to CompletionItemKind

### PR DESCRIPTION
**Pull request by Veykril**
Adds `LifetimeParam` and `ConstParam` to `CompletionItemKind` and maps them both to `TypeParam` in the protocol conversion as there are no equivalents, so nothing really changes there.
`ConstParam` could be mapped to `Const` I guess but I'm split on whether that would be better?

Additions were solely inspired by (the single) test output for const params.

Also sorts the variants of `CompletionItemKind` and its to_proto match.

### CODE REVIEW

**Comment by matklad** [comment]
Hm, let's apply  https://github.com/rust-analyzer/rust-analyzer/commit/c45221907adde640208a9e52636f4845d4654994 here as well. This requires movement of SymbolKind into ide_db

**Comment by Veykril** [comment]
Basically unifying it like this:
```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CompletionItemKind {
    SymbolKind(SymbolKind),
    Attribute,
    Binding,
    BuiltinType,
    Keyword,
    Method,
    Snippet,
    UnresolvedReference,
}
```
then, or fully unifying the two that only `SymbolKind` remains?

**Comment by matklad** [comment]
@Veykril I think more about the former. Snippet definitely dosn't belong to SymbolKind, although `BuiltinType` feels like it should be there.

File: crates/rust-analyzer/src/to_proto.rs [hunk 1]
```diff
@@ -91,24 +91,32 @@ pub(crate) fn completion_item_kind(
...
+            SymbolKind::SelfParam => lsp_types::CompletionItemKind::Value,
+            SymbolKind::Static => lsp_types::CompletionItemKind::Value,
+            SymbolKind::Struct => lsp_types::CompletionItemKind::Struct,
+            SymbolKind::Trait => lsp_types::CompletionItemKind::Interface,
+            SymbolKind::TypeAlias => lsp_types::CompletionItemKind::Struct,
+            SymbolKind::TypeParam => lsp_types::CompletionItemKind::TypeParameter,
+            SymbolKind::Union => lsp_types::CompletionItemKind::Struct,
+            SymbolKind::ValueParam => lsp_types::CompletionItemKind::Value,
+            SymbolKind::Variant => lsp_types::CompletionItemKind::EnumMember,
+        },
```
Comment (Veykril) [inline crates/rust-analyzer/src/to_proto.rs:119]: `SymbolKind::Impl => lsp_types::CompletionItemKind::Text,`
`SymbolKind::Label => lsp_types::CompletionItemKind::Variable,`
`SymbolKind::Local => lsp_types::CompletionItemKind::Variable,`
`SymbolKind::SelfParam => lsp_types::CompletionItemKind::Value,`
`SymbolKind::Union => lsp_types::CompletionItemKind::Struct,`
`SymbolKind::ValueParam => lsp_types::CompletionItemKind::Value,`
Are the newly added mappings here, I wasn't really sure what to map to what here so they might be very much wrong.

**Comment by Veykril** [comment]
`BuiltinType` doesn't have a corresponding entry in the protocols `SymbolKind` so I kept it in `CompletionItemKind` for the time being.

**Comment by matklad** [comment]
> BuiltinType doesn't have a corresponding entry in the protocols SymbolKind so
I kept it in CompletionItemKind for the time being.

At this level of abstraction, we don’t care in the least about LSP
protocol. We design for a hypothetical perfect client. So, we should do
what makes the most sense for rust. Lowering that to LSP is a problem of
to-proto module

**Comment by Veykril** [comment]
I guess I should've said instead that I'm not sure what to map that too as there is no `Type` Symbolkind or anything like that, only stuff like `struct`, `class`, `enum` etc. I guess struct should do then?

**Comment by matklad** [comment]
bors r+

I would't worry about precise mappings to LSP at this point. They are used solely to select an icon in VS Code, and code's icons are random anyhow: https://github.com/microsoft/vscode/issues/94263

**Pull request merged**